### PR TITLE
Initialize ipkernel at kernel creation time.

### PR DESCRIPTION
**Pull request by mdickinson**
[Includes commits from #229, which should be reviewed and merged first]

This PR updates the `InternalIPKernel` to initialise the kernel (i.e., execute the `init_ipkernel` method) at kernel creation time. Note that this means that `init_ipkernel` is called with `gui_backend = None`.

To allow other code to be adapted to this change, we make two more changes:

- it's (temporarily) permitted for `init_ipkernel` to be called twice on the same kernel, but doing so will issue a `DeprecationWarning`
- the single `gui_backend` argument to `init_ipkernel` has been made optional, and use of it is deprecated.

---

In addition to the changes above: the kernel initialization has been made (temporarily) opt-in. After the next release, it'll become the standard behaviour. This will give a chance for applications using this code to have a clean update path.

### CODE REVIEW

File: envisage/plugins/ipython_kernel/internal_ipkernel.py [hunk 1]
```diff
@@ -77,15 +78,37 @@ class InternalIPKernel(HasStrictTraits):
...
+            loop integration is set up.
+
+        .. note:: Use of this argument is deprecated!
         """
+        # For backwards compatibility, we allow a kernel to be initialised
+        # twice, and we ignore the second initialization, but warn.
+        if self.ipkernel is not None:
+            warnings.warn(
+                "IPython kernel is being initialised for a second time. "
+                "This may become an error in future versions of this package.",
```
Comment (rahulporuri) [inline envisage/plugins/ipython_kernel/internal_ipkernel.py]: Do you also want to mention/instead mention that the kernel has already been initialized and that does nothing?

It's not exactly clear from "IPython kernel is being initialised for a second time." that the `init_ipkernel` method doesn't do anything when called a second/nth time.

File: envisage/plugins/ipython_kernel/internal_ipkernel.py [hunk 2]
```diff
@@ -77,15 +78,37 @@ class InternalIPKernel(HasStrictTraits):
...
+            warnings.warn(
+                "The gui_backend argument is deprecated. "
+                "Integration with a GUI event loop can be "
+                "achieved by setting the 'eventloop' attribute on the "
+                "kernel instance",
+                DeprecationWarning,
+            )
+
         # Start IPython kernel with GUI event loop support
         self.ipkernel = gui_kernel(gui_backend)
```
Comment (rahulporuri) [inline envisage/plugins/ipython_kernel/internal_ipkernel.py]: does it make sense to also update `gui_kernel` to make `gui_backend` an optional argument?

Comment (mdickinson) [reply envisage/plugins/ipython_kernel/internal_ipkernel.py]: I don't think so; as far as I'm aware, no-one is using `gui_kernel` directly (but I'm happy to be corrected on that point).

File: envisage/plugins/ipython_kernel/internal_ipkernel.py (hunk 1 above)
Comment (mdickinson) [reply envisage/plugins/ipython_kernel/internal_ipkernel.py]: Yes, I'll update the message to clarify.

**Comment by mdickinson** [comment]
I was thinking about this overnight, and to avoid the unpleasantness of having to update Envisage simultaneously with its reverse dependencies, I think we need one more change here: a flag in the plugin that controls whether initialisation happens or not at startup time.

**Comment by mdickinson** [comment]
There's another issue here, which is that as written, if the service is not used, the `IPythonKernelPlugin` `stop` method instantiates and initialises the kernel just to shut it down. While we weren't initialising, that was less of an issue.

**Comment by mdickinson** [comment]
> if the service is not used, the `IPythonKernelPlugin` `stop` method instantiates and initialises the kernel just to shut it down

I'll fix this in a separate PR. It's a clear bug, so the fix can go into the next bugfix release of Envisage.

File: envisage/plugins/ipython_kernel/internal_ipkernel.py (hunk 1 above)
Comment (mdickinson) [reply envisage/plugins/ipython_kernel/internal_ipkernel.py]: Done.

**Comment by mdickinson** [comment]
@rahulporuri This has changed a lot since you last reviewed; sorry.

File: envisage/plugins/ipython_kernel/tests/test_ipython_kernel_plugin.py [hunk 3]
```diff
@@ -66,14 +67,112 @@ class NamespacePlugin(Plugin):
...
+                app.get_service(IPYTHON_KERNEL_PROTOCOL)
+
+        matching_messages = [
+            msg for msg in warn_msgs
+            if isinstance(msg.message, DeprecationWarning)
+            if "kernel will be initialized" in str(msg.message)
+        ]
+        self.assertEqual(len(matching_messages), 1)
+
+
```
Comment (mdickinson) [inline envisage/plugins/ipython_kernel/tests/test_ipython_kernel_plugin.py]: Excess blank lines; needs fixing.

File: envisage/plugins/ipython_kernel/internal_ipkernel.py (hunk 2 above)
Comment (mdickinson) [reply envisage/plugins/ipython_kernel/internal_ipkernel.py]: Though possibly we should make `gui_kernel` private in this PR.

Comment (mdickinson) [reply envisage/plugins/ipython_kernel/internal_ipkernel.py]: Done.

**Pull request merged**